File: sdk-kotlin/core/src/main/kotlin/ai/virion/capsule/core/Envelope.kt
// Provenance envelope: build, sign, verify.

package ai.virion.capsule.core

object Envelope {
    const val VERSION = "0.6"
    val SUPPORTED_CIPHERS = setOf("none", "ChaCha20-Poly1305")

    data class Signer(val role: String, val keyPair: CapsuleCrypto.Ed25519KeyPair)

    fun build(
        capsuleId: String,
        firstEventHash: String,
        entryHash: String,
        manifestHash: String,
        contentIndexHash: String,
        encryptedBlobHash: String? = null,
        cipher: String = "none",
        signedAt: String,
    ): JCSValue {
        require(cipher in SUPPORTED_CIPHERS) { "unsupported cipher: $cipher" }
        if (cipher == "none") require(encryptedBlobHash == null)
        else require(encryptedBlobHash?.length == 64)
        return JCSValue.Obj(listOf(
            "version" to JCSValue.Str(VERSION),
            "capsule_id" to JCSValue.Str(capsuleId),
            "first_event_hash" to JCSValue.Str(firstEventHash),
            "entry_hash" to JCSValue.Str(entryHash),
            "manifest_hash" to JCSValue.Str(manifestHash),
            "content_index_hash" to JCSValue.Str(contentIndexHash),
            "encrypted_blob_hash" to (encryptedBlobHash?.let { JCSValue.Str(it) } ?: JCSValue.Null),
            "cipher" to JCSValue.Str(cipher),
            "signed_at" to JCSValue.Str(signedAt),
            "signers" to JCSValue.Arr(emptyList()),
        ))
    }

    private fun canonicalPayload(envelope: JCSValue): ByteArray {
        val obj = envelope as JCSValue.Obj
        return JCS.bytes(JCSValue.Obj(obj.pairs.filterNot { it.first == "signers" }))
    }

    fun signingInput(envelope: JCSValue, role: String): ByteArray {
        require(role.isNotEmpty())
        val domain = "capsule-provenance-v$VERSION:$role ".toByteArray(Charsets.UTF_8)
        return CapsuleCrypto.concat(domain, canonicalPayload(envelope))
    }

    fun sign(envelope: JCSValue, signers: List<Signer>): JCSValue {
        val obj = envelope as JCSValue.Obj
        val sigIdx = obj.pairs.indexOfFirst { it.first == "signers" }
        require(sigIdx >= 0) { "envelope missing signers field" }
        val existing = obj.pairs[sigIdx].second as JCSValue.Arr
        require(existing.items.isEmpty()) { "envelope already has signers" }
        val signed = signers.map { s ->
            val input = signingInput(envelope, s.role)
            val sig = s.keyPair.sign(input)
            JCSValue.Obj(listOf(
                "role" to JCSValue.Str(s.role),
                "public_key" to JCSValue.Str(s.keyPair.publicKeyHex),
                "signature" to JCSValue.Str(CapsuleCrypto.bytesToHex(sig)),
            ))
        }
        val mutated = obj.pairs.toMutableList()
        mutated[sigIdx] = "signers" to JCSValue.Arr(signed)
        return JCSValue.Obj(mutated)
    }

    data class VerifyResult(
        val ok: Boolean,
        val signers: List<Triple<String, String, Boolean>>, // role, pkHex, valid
        val note: String? = null,
    )

    fun verifySignatures(envelope: JCSValue): VerifyResult {
        val obj = envelope as? JCSValue.Obj
            ?: return VerifyResult(false, emptyList(), "envelope is not an object")
        val versionStr = (obj.pairs.firstOrNull { it.first == "version" }?.second as? JCSValue.Str)?.v
        if (versionStr != VERSION) return VerifyResult(false, emptyList(), "unsupported version")
        val cipher = (obj.pairs.firstOrNull { it.first == "cipher" }?.second as? JCSValue.Str)?.v
        if (cipher !in SUPPORTED_CIPHERS) return VerifyResult(false, emptyList(), "unsupported cipher")
        val signers = (obj.pairs.firstOrNull { it.first == "signers" }?.second as? JCSValue.Arr)?.items
            ?: return VerifyResult(false, emptyList(), "no signers")
        if (signers.isEmpty()) return VerifyResult(false, emptyList(), "envelope has no signers")
        var allValid = true
        val results = mutableListOf<Triple<String, String, Boolean>>()
        for (s in signers) {
            val sObj = s as? JCSValue.Obj ?: run { allValid = false; continue }
            val role = (sObj.pairs.firstOrNull { it.first == "role" }?.second as? JCSValue.Str)?.v
            val pk = (sObj.pairs.firstOrNull { it.first == "public_key" }?.second as? JCSValue.Str)?.v
            val sig = (sObj.pairs.firstOrNull { it.first == "signature" }?.second as? JCSValue.Str)?.v
            if (role == null || pk == null || sig == null) { allValid = false; continue }
            val input = signingInput(envelope, role)
            val valid = CapsuleCrypto.ed25519Verify(
                CapsuleCrypto.hexToBytes(pk), input, CapsuleCrypto.hexToBytes(sig)
            )
            if (!valid) allValid = false
            results += Triple(role, pk, valid)
        }
        return VerifyResult(allValid, results)
    }
}

File: sdk-swift/Sources/Capsule/Envelope.swift
// Provenance envelope: build, sign, verify. Mirrors envelope.js.

import Foundation

public enum Envelope {
    static let VERSION = "0.6"
    static let SUPPORTED_CIPHERS: Set<String> = ["none", "ChaCha20-Poly1305"]

    public struct Signer {
        public let role: String
        public let keyPair: Ed25519KeyPair
        public init(role: String, keyPair: Ed25519KeyPair) {
            self.role = role; self.keyPair = keyPair
        }
    }

    public static func build(
        capsuleId: String,
        firstEventHash: String,
        entryHash: String,
        manifestHash: String,
        contentIndexHash: String,
        encryptedBlobHash: String? = nil,
        cipher: String = "none",
        signedAt: String
    ) -> JCSValue {
        precondition(SUPPORTED_CIPHERS.contains(cipher), "unsupported cipher")
        if cipher == "none" {
            precondition(encryptedBlobHash == nil, "plain capsule must have encrypted_blob_hash=null")
        } else {
            precondition(encryptedBlobHash != nil, "encrypted capsule needs blob hash")
        }
        return .object([
            ("version", .string(VERSION)),
            ("capsule_id", .string(capsuleId)),
            ("first_event_hash", .string(firstEventHash)),
            ("entry_hash", .string(entryHash)),
            ("manifest_hash", .string(manifestHash)),
            ("content_index_hash", .string(contentIndexHash)),
            ("encrypted_blob_hash", encryptedBlobHash.map { .string($0) } ?? .null),
            ("cipher", .string(cipher)),
            ("signed_at", .string(signedAt)),
            ("signers", .array([])),
        ])
    }

    /// Returns JCS bytes of the envelope minus the `signers` field.
    static func canonicalPayload(_ envelope: JCSValue) -> Data {
        guard case .object(let pairs) = envelope else {
            preconditionFailure("envelope is not an object")
        }
        let withoutSigners = pairs.filter { $0.0 != "signers" }
        return JCS.bytes(.object(withoutSigners))
    }

    /// `domain_sep || canonical(envelope_minus_signers)` — the signing input.
    static func signingInput(_ envelope: JCSValue, role: String) -> Data {
        precondition(!role.isEmpty)
        let domain = Data("capsule-provenance-v\(VERSION):\(role)\0".utf8)
        return Bytes.concat(domain, canonicalPayload(envelope))
    }

    public static func sign(_ envelope: inout JCSValue, signers: [Signer]) throws {
        guard case .object(var pairs) = envelope else {
            preconditionFailure("envelope is not an object")
        }
        guard let sIdx = pairs.firstIndex(where: { $0.0 == "signers" }),
              case .array(let existing) = pairs[sIdx].1, existing.isEmpty
        else { throw CapsuleError.malformed("envelope already has signers") }

        var signed: [JCSValue] = []
        for s in signers {
            let input = signingInput(envelope, role: s.role)
            let sig = try s.keyPair.sign(input)
            signed.append(.object([
                ("role", .string(s.role)),
                ("public_key", .string(s.keyPair.publicKeyHex)),
                ("signature", .string(Bytes.toHex(sig))),
            ]))
        }
        pairs[sIdx] = ("signers", .array(signed))
        envelope = .object(pairs)
    }

    public struct VerifyResult {
        public let ok: Bool
        public let signers: [(role: String, publicKey: String, valid: Bool)]
        public let note: String?
    }

    public static func verifySignatures(_ envelope: JCSValue) -> VerifyResult {
        guard case .object(let pairs) = envelope,
              let versionPair = pairs.first(where: { $0.0 == "version" }),
              case .string(let v) = versionPair.1, v == VERSION
        else { return VerifyResult(ok: false, signers: [], note: "unsupported version") }

        guard let cipherPair = pairs.first(where: { $0.0 == "cipher" }),
              case .string(let c) = cipherPair.1, SUPPORTED_CIPHERS.contains(c)
        else { return VerifyResult(ok: false, signers: [], note: "unsupported cipher") }

        guard let signersPair = pairs.first(where: { $0.0 == "signers" }),
              case .array(let signersArr) = signersPair.1, !signersArr.isEmpty
        else { return VerifyResult(ok: false, signers: [], note: "no signers") }

        var allValid = true
        var out: [(String, String, Bool)] = []
        for s in signersArr {
            guard case .object(let sp) = s,
                  let rolePair = sp.first(where: { $0.0 == "role" }),
                  case .string(let role) = rolePair.1,
                  let pkPair = sp.first(where: { $0.0 == "public_key" }),
                  case .string(let pkHex) = pkPair.1,
                  let sigPair = sp.first(where: { $0.0 == "signature" }),
                  case .string(let sigHex) = sigPair.1
            else { allValid = false; continue }
            // Hex strings on the wire are caller-supplied — never panic.
            // A malformed signer is just an invalid signature.
            guard let pkBytes = try? Bytes.fromHexThrowing(pkHex, label: "signer.public_key"),
                  let sigBytes = try? Bytes.fromHexThrowing(sigHex, label: "signer.signature")
            else {
                allValid = false
                out.append((role, pkHex, false))
                continue
            }
            let input = signingInput(envelope, role: role)
            let valid = Ed25519.verify(
                publicKey: pkBytes,
                message: input,
                signature: sigBytes
            )
            if !valid { allValid = false }
            out.append((role, pkHex, valid))
        }
        return VerifyResult(ok: allValid, signers: out, note: nil)
    }
}
